File: app/app/src/main/java/com/shelter/shade/engine/Routing.kt
package com.shelter.shade.engine

import java.util.PriorityQueue
import kotlin.math.tan

/** OSM 보행 그래프 기반 그늘 가중 라우팅. shade-engine/shade_engine/osm_routing.py 이식. */

data class RouteOption(
    val name: String,
    val alpha: Double,
    val coords: List<DoubleArray>, // [[lat,lon],...]
    val distanceM: Double,
    val sunFraction: Double,
) {
    val shadePercent: Double get() = ((1.0 - sunFraction) * 1000).toInt() / 10.0
}

class RouteNotFoundException(message: String) : RuntimeException(message)

private const val MAX_SHADOW_DISTANCE_CAP_M = 1500.0
private const val DEFAULT_MAX_SNAP_M = 200.0

fun planRoutesOsm(
    graph: OsmGraph,
    origin: DoubleArray, // [lat,lon]
    dest: DoubleArray,
    buildings: List<Building>,
    sunAzimuthDeg: Double,
    sunAltitudeDeg: Double,
    alphas: Map<String, Double>? = null,
    preferSun: Boolean = false,
    maxSnapM: Double = DEFAULT_MAX_SNAP_M,
): List<RouteOption> {
    val effectiveAlphas = alphas ?: linkedMapOf(
        "shortest" to 0.0,
        "balanced" to 3.0,
        (if (preferSun) "sunniest" else "shadiest") to 12.0,
    )

    if (graph.nodeCount() == 0) throw RouteNotFoundException("빈 보행 그래프")

    val lat0 = (origin[0] + dest[0]) / 2
    val lon0 = (origin[1] + dest[1]) / 2
    val proj = LocalProjection(lat0, lon0)
    val projected = projectBuildings(buildings, proj)
    val tallest = projected.maxOfOrNull { it.heightM } ?: 0.0
    val tanAlt = tan(Math.toRadians(maxOf(sunAltitudeDeg, 0.0)))
    val maxDist = if (tanAlt > 1e-6) minOf(MAX_SHADOW_DISTANCE_CAP_M, tallest / tanAlt) else MAX_SHADOW_DISTANCE_CAP_M
    val index = BuildingIndex(projected)

    val n = graph.nodeCount()
    val sunny = BooleanArray(n)
    val blocked = BooleanArray(n)
    for (i in 0 until n) {
        val xy = proj.toXy(graph.nodes[i][0], graph.nodes[i][1])
        val res = index.isPointShaded(xy, sunAzimuthDeg, sunAltitudeDeg, maxDistanceM = maxOf(maxDist, 20.0))
        sunny[i] = !res.shaded
        blocked[i] = res.reason == "inside_building"
    }

    val start = graph.nearestNode(origin[0], origin[1])
    val goal = graph.nearestNode(dest[0], dest[1])
    val oSnap = haversineM(origin[0], origin[1], graph.nodes[start][0], graph.nodes[start][1])
    val dSnap = haversineM(dest[0], dest[1], graph.nodes[goal][0], graph.nodes[goal][1])
    if (oSnap > maxSnapM || dSnap > maxSnapM) {
        throw RouteNotFoundException("보행망 범위 밖(스냅 ${oSnap.toInt()}m/${dSnap.toInt()}m)")
    }
    if (start == goal && haversineM(origin[0], origin[1], dest[0], dest[1]) > maxSnapM) {
        throw RouteNotFoundException("출발/도착이 같은 노드로 스냅됨")
    }
    blocked[start] = false
    blocked[goal] = false

    val avoid = DoubleArray(n) { i ->
        val a = if (preferSun) !sunny[i] else sunny[i]
        if (a) 1.0 else 0.0
    }

    val options = ArrayList<RouteOption>()
    for ((name, alpha) in effectiveAlphas) {
        val path = dijkstra(graph, start, goal, avoid, blocked, alpha)
            ?: throw RouteNotFoundException("출발-도착이 연결되지 않음")
        val coords = ArrayList<DoubleArray>(maxOf(path.size, 2))
        if (path.size <= 1) {
            // 출발·도착이 같은 노드로 스냅된 짧은 경로 → 양 끝점을 모두 보존
            coords.add(doubleArrayOf(origin[0], origin[1]))
            coords.add(doubleArrayOf(dest[0], dest[1]))
        } else {
            for (idx in path) coords.add(doubleArrayOf(graph.nodes[idx][0], graph.nodes[idx][1]))
            coords[0] = doubleArrayOf(origin[0], origin[1])
            coords[coords.size - 1] = doubleArrayOf(dest[0], dest[1])
        }
        var dist = 0.0
        for (k in 0 until coords.size - 1) {
            dist += haversineM(coords[k][0], coords[k][1], coords[k + 1][0], coords[k + 1][1])
        }
        val sunNodes = path.count { sunny[it] }
        val sunFrac = if (path.isEmpty()) 0.0 else sunNodes.toDouble() / path.size
        options.add(RouteOption(name, alpha, coords, dist, sunFrac))
    }
    return options
}

private fun dijkstra(
    graph: OsmGraph,
    start: Int,
    goal: Int,
    avoid: DoubleArray,
    blocked: BooleanArray,
    alpha: Double,
): List<Int>? {
    val dist = HashMap<Int, Double>()
    val prev = HashMap<Int, Int>()
    dist[start] = 0.0
    val pq = PriorityQueue<DoubleArray>(compareBy { it[0] }) // [cost, node]
    pq.add(doubleArrayOf(0.0, start.toDouble()))
    val visited = HashSet<Int>()

    while (pq.isNotEmpty()) {
        val top = pq.poll() ?: break
        val d = top[0]
        val u = top[1].toInt()
        if (u == goal) break
        if (u in visited) continue
        visited.add(u)
        for (edge in graph.adj[u].orEmpty()) {
            val v = edge.to
            if (blocked[v] && v != goal) continue
            val edgeAvoid = (avoid[u] + avoid[v]) / 2.0
            val nd = d + edge.lenM * (1.0 + alpha * edgeAvoid)
            if (nd < (dist[v] ?: Double.MAX_VALUE)) {
                dist[v] = nd
                prev[v] = u
                pq.add(doubleArrayOf(nd, v.toDouble()))
            }
        }
    }

    if (goal !in dist) return null
    val path = ArrayList<Int>()
    var cur = goal
    path.add(cur)
    while (cur != start) {
        cur = prev[cur] ?: return null
        path.add(cur)
    }
    path.reverse()
    return path
}

File: shade-engine/shade_engine/routing.py
"""그늘 가중 라우팅 (Phase 2).

Valhalla 커스텀 코스팅의 핵심 원리를 격자 그래프 + 다익스트라로 구현한다.
엣지 비용 = 거리 × (1 + α · 햇빛비율). α 를 키우면 '가장 그늘진 길', 0 이면 '최단'.
실제 운영에서는 OSM 도로 그래프 + Valhalla 로 대체하되, 가중치 공식은 동일하다.

여기서는 좁은 권역(출발~도착 bbox)에서 8방향 격자를 만들고, 각 노드의 햇빛 여부를
태양 1회 계산(좁은 영역이라 거의 일정)으로 구한 뒤 경로를 탐색한다.
"""

from __future__ import annotations

import heapq
import math
from dataclasses import dataclass

from .buildings import Building
from .geo import LocalProjection, haversine_m
from .raycast import ProjectedBuilding, is_point_shaded

_MAX_SHADOW_DISTANCE_CAP_M = 1500.0


@dataclass(frozen=True)
class RouteOption:
    """라우팅 결과 하나."""

    name: str  # 'shortest' | 'balanced' | 'shadiest'
    alpha: float
    coords: list[tuple[float, float]]  # [(lat, lon), ...]
    distance_m: float
    sun_fraction: float  # 경로 노드 중 햇빛 비율 0~1

    @property
    def shade_percent(self) -> float:
        return round((1.0 - self.sun_fraction) * 100.0, 1)


def _project_buildings(buildings: list[Building], proj: LocalProjection) -> list[ProjectedBuilding]:
    out: list[ProjectedBuilding] = []
    for b in buildings:
        ring_xy = tuple(proj.to_xy(lat, lon) for lat, lon in b.ring)
        if len(ring_xy) >= 3:
            out.append(
                ProjectedBuilding(
                    ring_xy=ring_xy,
                    height_m=b.height_m,
                    height_estimated=b.height_estimated,
                    osm_id=b.osm_id,
                )
            )
    return out


def plan_routes(
    origin: tuple[float, float],
    dest: tuple[float, float],
    buildings: list[Building],
    sun_azimuth_deg: float,
    sun_altitude_deg: float,
    *,
    grid_spacing_m: float = 20.0,
    margin_m: float = 150.0,
    max_nodes: int = 6000,
    alphas: dict[str, float] | None = None,
    prefer_sun: bool = False,
) -> list[RouteOption]:
    """origin→dest 의 최단/균형/그늘(또는 햇빛) 경로 후보를 만든다.

    grid_spacing_m 간격의 8방향 격자에서 알파별 다익스트라를 돌린다. 격자가 너무
    커지면(max_nodes 초과) 간격을 자동으로 키워 노드 수를 제한한다. prefer_sun=True
    (겨울 모드)면 그늘 대신 햇빛을 최대화한다.
    """
    if alphas is None:
        third = "sunniest" if prefer_sun else "shadiest"
        alphas = {"shortest": 0.0, "balanced": 3.0, third: 12.0}

    (lat0, lon0), (lat1, lon1) = origin, dest
    min_lat, max_lat = sorted((lat0, lat1))
    min_lon, max_lon = sorted((lon0, lon1))

    proj = LocalProjection(lat0=(min_lat + max_lat) / 2, lon0=(min_lon + max_lon) / 2)
    ox, oy = proj.to_xy(*origin)
    dx, dy = proj.to_xy(*dest)
    x_lo, x_hi = sorted((ox, dx))
    y_lo, y_hi = sorted((oy, dy))
    x_lo -= margin_m
    x_hi += margin_m
    y_lo -= margin_m
    y_hi += margin_m

    # 노드 수 제한을 위해 필요 시 간격 자동 확대
    spacing = grid_spacing_m
    while ((x_hi - x_lo) / spacing + 1) * ((y_hi - y_lo) / spacing + 1) > max_nodes:
        spacing *= 1.5

    ncols = int((x_hi - x_lo) / spacing) + 1
    nrows = int((y_hi - y_lo) / spacing) + 1

    def node_xy(r: int, c: int) -> tuple[float, float]:
        return (x_lo + c * spacing, y_lo + r * spacing)

    def snap(x: float, y: float) -> tuple[int, int]:
        c = min(ncols - 1, max(0, round((x - x_lo) / spacing)))
        r = min(nrows - 1, max(0, round((y - y_lo) / spacing)))
        return (r, c)

    projected = _project_buildings(buildings, proj)
    tallest = max((b.height_m for b in projected), default=0.0)
    tan_alt = math.tan(math.radians(max(sun_altitude_deg, 0.0)))
    max_dist = (
        min(_MAX_SHADOW_DISTANCE_CAP_M, tallest / tan_alt) if tan_alt > 1e-6 else _MAX_SHADOW_DISTANCE_CAP_M
    )

    # 각 노드의 햇빛 여부 + 통행 가능 여부를 1회 계산(좁은 영역이라 태양 위치는 거의 일정).
    # 건물 내부 노드는 통행 불가(blocked) — 그렇지 않으면 '값싼 그늘'로 오인되어
    # 경로가 건물을 관통할 수 있다.
    sunny = [[False] * ncols for _ in range(nrows)]
    blocked = [[False] * ncols for _ in range(nrows)]
    for r in range(nrows):
        for c in range(ncols):
            res = is_point_shaded(
                node_xy(r, c),
                sun_azimuth_deg,
                sun_altitude_deg,
                projected,
                max_distance_m=max(max_dist, spacing),
            )
            sunny[r][c] = not res.shaded
            blocked[r][c] = res.reason == "inside_building"

    start = snap(ox, oy)
    goal = snap(dx, dy)
    # 출발/도착이 건물 내부로 스냅되어도 경로가 성립하도록 통행 허용
    blocked[start[0]][start[1]] = False
    blocked[goal[0]][goal[1]] = False

    # 회피 대상: 여름은 햇빛, 겨울(prefer_sun)은 그늘
    avoid = (
        [[not sunny[r][c] for c in range(ncols)] for r in range(nrows)] if prefer_sun else sunny
    )

    options: list[RouteOption] = []
    for name, alpha in alphas.items():
        path_nodes = _dijkstra(start, goal, nrows, ncols, spacing, avoid, blocked, alpha)
        coords = [proj.to_latlon(*node_xy(r, c)) for r, c in path_nodes]
        # 정확한 출발/도착 좌표로 끝점 치환(격자 스냅 오차 보정)
        if coords:
            coords[0] = origin
            coords[-1] = dest
        dist = sum(
            haversine_m(coords[i][0], coords[i][1], coords[i + 1][0], coords[i + 1][1])
            for i in range(len(coords) - 1)
        )
        sun_nodes = sum(1 for r, c in path_nodes if sunny[r][c])
        sun_frac = sun_nodes / len(path_nodes) if path_nodes else 0.0
        options.append(
            RouteOption(name=name, alpha=alpha, coords=coords, distance_m=dist, sun_fraction=sun_frac)
        )
    return options


# 8방향 이웃 (dr, dc)
_NEIGHBORS = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]


def _dijkstra(
    start: tuple[int, int],
    goal: tuple[int, int],
    nrows: int,
    ncols: int,
    spacing: float,
    avoid: list[list[bool]],
    blocked: list[list[bool]],
    alpha: float,
) -> list[tuple[int, int]]:
    """엣지비용 = 거리 × (1 + α · 엣지회피비율) 다익스트라. 경로 노드열 반환.

    avoid 는 회피 대상(여름=햇빛, 겨울=그늘) 노드. blocked 노드(건물 내부)는 통행 불가.
    """
    def idx(r: int, c: int) -> int:
        return r * ncols + c

    dist = [math.inf] * (nrows * ncols)
    prev: list[int] = [-1] * (nrows * ncols)
    sr, sc = start
    dist[idx(sr, sc)] = 0.0
    pq: list[tuple[float, int, int]] = [(0.0, sr, sc)]

    while pq:
        d, r, c = heapq.heappop(pq)
        if (r, c) == goal:
            break
        if d > dist[idx(r, c)]:
            continue
        for dr, dc in _NEIGHBORS:
            nr, nc = r + dr, c + dc
            if not (0 <= nr < nrows and 0 <= nc < ncols):
                continue
            if blocked[nr][nc]:
                continue  # 건물 내부 통행 불가
            step = spacing * (math.sqrt(2) if dr and dc else 1.0)
            # 엣지 회피비율 = 양 끝 노드 평균
            edge_avoid = (int(avoid[r][c]) + int(avoid[nr][nc])) / 2.0
            cost = step * (1.0 + alpha * edge_avoid)
            nd = d + cost
            if nd < dist[idx(nr, nc)]:
                dist[idx(nr, nc)] = nd
                prev[idx(nr, nc)] = idx(r, c)
                heapq.heappush(pq, (nd, nr, nc))

    # 경로 복원
    gi = idx(*goal)
    if dist[gi] == math.inf:
        return [start, goal]
    path: list[tuple[int, int]] = []
    cur = gi
    while cur != -1:
        path.append((cur // ncols, cur % ncols))
        cur = prev[cur]
    path.reverse()
    return path
